src/item.h

```diff
@@ -206,6 +207,226 @@ class ItemAttributes
 			return static_cast<ItemDecayState_t>(getIntAttr(ITEM_ATTRIBUTE_DECAYSTATE));
 		}
 
+		struct CustomAttributeKey
+		{
+			union {
+				std::string* string;
+				int64_t integer;
+			} value;
+
+			customAttrTypes type = ATTR_NO_TYPE;
+
+			explicit CustomAttributeKey(customAttrTypes type) : type(type) {
+				memset(&value, 0, sizeof(value));
+			}
+
+			CustomAttributeKey(const std::string& v) {
+				type = ATTR_STRING_TYPE;
+				value.string = new std::string(v);
+			}
+
+			CustomAttributeKey(const char* v) {
+				type = ATTR_STRING_TYPE;
+				value.string = new std::string(v);
+			}
+
+			CustomAttributeKey(int64_t v) {
+				type = ATTR_INTEGER_TYPE;
+				value.integer = v;
+			}
+
+			CustomAttributeKey(const CustomAttributeKey& i) {
+				type = i.type;
+				switch (type) {
+					case ATTR_STRING_TYPE: {
+						value.string = new std::string(*i.value.string);
+					}
+
+					case ATTR_INTEGER_TYPE: {
+						value.integer = i.value.integer;
+					}
+
+					default: {
+						memset(&value, 0, sizeof(value));
+					}
+				}
+			}
+
+			CustomAttributeKey(CustomAttributeKey&& attribute) : value(attribute.value), type(attribute.type) {
+				memset(&attribute.value, 0, sizeof(value));
```

Comment:
memset/memcpy is considered bad practice in C++11 and newer (refer [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#a-namepro-bounds-stdlibabounds4-dont-use-standard-library-functions-and-types-that-are-not-bounds-checked)).